### DISCUSSION TITLE
ValueError: Short orders require: TP < LIMIT < SL

### DISCUSSION THREAD

**Discussion by CadeHalcyon** [kernc/backtesting.py#850] [General]
Hi, sometimes I get the following error code when running my strategy, same can happen with long/buy orders (something like ValueError:  Buy orders require TP>LIMIT>SL).  I understand what this means, but not _why_ it is happening and therefore am stumped how to resolve it.  Probably something simple - grateful for any help.  Thanks.

`ValueError: Short orders require: TP (1.262196) < LIMIT (1.1701024400000002) < SL (1.121952)`

Here's the code causing the problem in this particular instance:

```
class Strategy(Strategy):        
    stop_loss = 0.04
    take_profit = 0.08
    max_size = 50000
    equity_size = 0.25 
    leverage = 25
    indicator_3_value = 15

 def init(self):
    self.EMA = self.I(lambda EMA: EMA, self.data.EMA.s)
    self.indicator_1 = self.I(lambda indicator_1: indicator_1, self.data.indicator_1.s)
    self.indicator_2 = self.I(lambda indicator_2: indicator_2, self.data.indicator_2.s)
    self.indicator_3 = self.I(lambda indicator_3: indicator_3, self.data.indicator_3.s
 
def next(self):
    price = self.data.Close[-1]
    self._trade_equity = self.equity * self.equity_size * self.leverage
    trade_size = min(self.max_size, (self._trade_equity // price))
    
    if not self.position: 
         if price<self.EMA[-1] and crossover(self.indicator_1, self.indicator_2) and self.indicator_3[-1]>self.indicator_3_value: 
                    self.sell(sl=(1+self.stop_loss)*price,tp=(1-self.take_profit)*price,size=trade_size)

bt = Backtest(df_2, Strategy, cash=1000, 
              commission=0.0012, exclusive_orders=True, trade_on_close = True, margin=0.04)

stats = bt.optimize(stop_loss = np.arange(0.01, 0.05, 0.005).tolist(),
                    take_profit = np.arange(0.015,0.10,0.005).tolist(),
                    maximize='Return [%]', 
                    method="grid",return_heatmap=True,
                    )
```

**Comment by kernc** [comment]
Your SL/TP need to be more than `commission` away from your market order/limit price. In reality, a broker cancels such an order.
See https://github.com/kernc/backtesting.py/discussions/605#discussioncomment-2247851.

**Reply by CadeHalcyon** [reply]
Thanks but my commission is set to 0.0012, and lowest stop_loss and take_profit are 0.01 and 0.015 respectively for the bt.optimize.  I assumed that the entry price would be self.data.Close[-1].  Is this not the case?  How do I reference the entry price for the SL and TP please, per the "?????" below?  

`self.sell(sl=(1+self.stop_loss)*?????, tp=(1-self.take_profit)*?????, size=trade_size)`

**Reply by CadeHalcyon** [reply]
I've also tried removing commission and still get the same issue

**Reply by CadeHalcyon** [reply]
I seem to have resolved this by replacing the SL and TP in `self.sell(sl=(1+self.stop_loss)*price,tp=(1-self.take_profit)*price,size=trade_size)` with the following - though the SLs and TPs are rarely 0.04 and 0.08 even accounting for commission, which makes sense seeing that the close will never be exact (though when I set SLs and TPs using self.sell and self.buy they were pretty much the exact %s +/- commission):

```
self.sell(size=trade_size)    
self.buy(size=trade_size)

        if self.position:
               for trade in self.trades:
                    if trade.is_long:
                        if self.data.Close[-1] <= trade.entry_price * (1-self.stop_loss):
                            trade.close()
                        if self.data.Close[-1] >= trade.entry_price * (1+self.take_profit):
                            trade.close()
                    elif trade.is_short:
                        if self.data.Close[-1] >= trade.entry_price * (1+self.stop_loss):
                            trade.close()
                        if self.data.Close[-1] <= trade.entry_price * (1-self.take_profit):
                            trade.close()
```
